### ISSUE TITLE
Support for Symfony 6 + DBAL3 (tests), and PHP 8.1 support 

### ISSUE THREAD

**Issue by yivi** [DamienHarper/auditor#68]
TLDR; Attempting to create a PR to add support for Symfony 6, but many issues crop.

Would it be worth attempting to fix these issues on branch 1.x, and have support from Sf 3 to Sf 6 on this branch... or would this efforts better directed at the v2? Branch? Do you consider v2 branch to be near enough to be released in the near future? 

Now, the long version:

| Q                    | A
| ---------------------| -----------------------
| `auditor` version    | 1.3.1
| PHP version          | >= 8

Right now the package is not installable/runnable on Symfony 6.

I've created a branch to see exactly what's blocking the upgrade.

First, getting the obvious out of the way, on `composer.json` we have:

```
    "symfony/event-dispatcher": "^3.4|^4.0|^5.0",
    "symfony/lock": "^3.4|^4.0|^5.0",
    "symfony/options-resolver": "^3.4|^4.0|^5.0"
```
and on `require-dev`:
```
    "symfony/var-dumper": "^4.0|^5.0"
```

Adding`"|^6.0"` everywhere would make it installable on Sf 6 at least (with `--no-dev`), but only for things to break elsewhere.

First, the less important issue (as I see it).

The use of `"friendsofphp/php-cs-fixer": "^2.16`. This version is only compatible with Symfony 3, 4, or 5. **But not 6**.  And PHP-CS-Fixer ^3,it's not compatible with Symfony ^3. (First version with Sf 6 support is 3.4). So the same version of PHP-CS-Fixer cannot be installed for versions 3 to 6 of Symfony. **And** PHP-CS-Fixer config for v2 is not compatible with PHP-CS-Fixer v3...

This is bothersome because this is not even a "real" dependency, but simply an external tool, without any runtime impact. In real life, this shouldn't even be a compatibility issue. But it is because of installing dev tools via composer.

This hurdle _could_ be sidestepped. A couple of options:

* Install `friendsofphp/php-cs-fixer` as a PHAR e.g. using something like [this](https://pharhub.dev/), within composer; or like [this](https://phar.io/), outisde
* Install `friendsofphp/php-cs-fixer` using something like [this](https://github.com/bamarni/composer-bin-plugin), so it's installed on its separate  `vendor-bin` directory.

Considering the nature of the tool, any option should be better than allowing it to trigger dependency hell.

But even with this out of the way:

The project requires `"doctrine/doctrine-migrations-bundle": "^1.3|^2.0"`, but only `doctrine-migrations-bundle >= 3` declare support for Sf 6. So we could change this to `"^1.3|^2.0|^3.0"` to allow `--dev` installation...

If that's the case, on Sf 6 doctrine-migrations-bundle v^3 would be installed, and at least _this_ wouldn't block the installation.

But without adding a `conflict` rule for `dbal >= 3`, DBAL v 3 would be installed, and this would case **several** problems, with many, as far as I can see only on happening on `--dev`, since many tests break under DBAL 3 (the library itself works **fine** with DBAL 3, as far as I can see). So declaring the library to be incompatible with seems not great.

There is quite a bit to go to make this compatible with a Symfony `^3.4|^4.4|^5.4|^6.0`...

Dropping v3 would alleviate this a fair bit... but could be an unacceptable compromise.

Before I keep digging and trying to see if a PR would make sense, what do you think of this? Is dropping v3 a no-go? Is adding v6 (and DBAL 3 for tests) support too much for this version of the library? Would you welcome efforts on this, or rather have contributions focus on the future v2? 

Hopefully I can lend a hand in some capacity.

And happy new year! :)

**Comment by DamienHarper** [comment]
@yivi Thanks for trying to bring sf6 compat to this lib, much appreciated 👍🏽 
Here is my point on your first questions:
> Would it be worth attempting to fix these issues on branch 1.x, and have support from Sf 3 to Sf 6 on this branch... or would this efforts better directed at the v2 Branch? 

As I mentioned in the `README.md` on `master` branch (which is the v2 branch), v2 drops support of sf3.4 and PHP 7.2 (still thinking about dropping PHP 7.3 support as well as it is not supported anymore for quite a long time now). 
`1.x` branch will only receive critical and/or security fixes so using `master` branch (v2) it is the way to go.

> Do you consider v2 branch to be near enough to be released in the near future?

I hope so, I'd like to release v2 sometime in late January or February (no fixed date though).

Regarding external tools such as PHPStan and PHP-CS-Fixer, I know they become a blocker and should not belong to real dev dependencies so we have to find a way to workaround this. Right now, I do not have a preferred way of doing this but having them installed in a dedicated `composer.json` file (and dedicated bin folder) seems a good start.

Best wishes for this new year!

**Comment by yivi** [comment]
Great! Are there any tasks with which one could help? Any "help-wanted-accepted" issues where a PR would be welcome on the `master` branch?

**Comment by DamienHarper** [comment]
@yivi I haven't tested sf6 compatibility, any help on this would be much appreciated :)

**Comment by yivi** [comment]
After setting the the symfony/* deps to `^6.0` and updating, all tests run fine without issue. I'll make more checks on the "bundle" package later.

What PHP version are you targeting for this release? On the readme it says >= 7.4, but I've seen a couple `mixed` return type-hint that would break on PHP < 8 (twice in `AuditTrait`, and twice in `RangeFilter`). Is that correct?

I'm fine with >= 8 if that's the case.

On another note, great thing moving the tools to `tools`. I would personally include the lock-files for each tool as well, so anyone testing the tool would use the exact same version of the tool than you or CI.

**Comment by DamienHarper** [comment]
@yivi 

> After setting the the symfony/* deps to `^6.0` and updating, all tests run fine without issue. I'll make more checks on the "bundle" package later.

Great, thanks!
 
> What PHP version are you targeting for this release? On the readme it says >= 7.4, but I've seen a couple `mixed` return type-hint that would break on PHP < 8 (twice in `AuditTrait`, and twice in `RangeFilter`). Is that correct?
>
> I'm fine with >= 8 if that's the case.
> 

As mentioned in the `README.md` I target PHP >= 7.4 so those `mixed` returns should be removed, you're right.

> On another note, great thing moving the tools to `tools`. I would personally include the lock-files for each tool as well, so anyone testing the tool would use the exact same version of the tool than you or CI.

Yeah, this has to be done too ;)
